From c80dd077fcb82d4213fff8ccf3c69da2058bb894 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Thu, 19 Nov 2020 18:00:48 +0100
Subject: [PATCH] PONRTSYS-8285: Use pskb_network_may_pull() in IPv4 hook

Use the pskb_network_may_pull() function to check if the skb is big
enough and then also use sizeof(struct iphdr) instead of hard coding
the 5.
---
 .../mcast_bridge_filter/mcast_bridge_filter.c      | 41 +++++-----------------
 1 file changed, 8 insertions(+), 33 deletions(-)

diff --git a/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c b/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
index 002578757b85..f5a6682245af 100644
--- a/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
+++ b/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
@@ -15,6 +15,7 @@
 #include <linux/if_vlan.h>
 #include <linux/in.h>
 #include <linux/ip.h>
+#include <net/addrconf.h>
 #include <net/mcast_bridge_filter.h>
 
 /* max number of encapsulated vlan tags */
@@ -51,38 +52,15 @@ static int dp_mcast_vlan_iter(struct sk_buff *skb, __be16 *proto, int *nhoff)
 	return 0;
 }
 
-/* Returns 1 for an IPv6 MLD or MGM packet, 0, -EINVAL on an parsing error and
- * 0 if it is not an IPv6 MLD or MGM packet.
- */
-static int dp_mcast_ipv6_is_mld_pkt(struct sk_buff *skb, int nhoff)
-{
-	const struct icmp6hdr *icmp6hdr;
-	struct icmp6hdr _icmp6hdr;
-
-	icmp6hdr = skb_header_pointer(skb, nhoff, sizeof(_icmp6hdr),
-				      &_icmp6hdr);
-	if (!icmp6hdr)
-		return -EINVAL;
-
-	switch (icmp6hdr->icmp6_type) {
-	case ICMPV6_MGM_QUERY:
-	case ICMPV6_MGM_REPORT:
-	case ICMPV6_MGM_REDUCTION:
-	case ICMPV6_MLD2_REPORT:
-		return 1;
-	default:
-		break;
-	}
-	return 0;
-}
-
 static unsigned int dp_mcast_br_hook_ipv4(struct sk_buff *skb, int nhoff)
 {
-	const struct iphdr *iph;
-	struct iphdr _iph;
+	struct iphdr *iph;
 
-	iph = skb_header_pointer(skb, nhoff, sizeof(_iph), &_iph);
-	if (!iph || iph->ihl < 5) {
+	if (!pskb_network_may_pull(skb, nhoff + sizeof(struct iphdr)))
+		return NF_ACCEPT;
+
+	iph = (struct iphdr *)(skb_network_header(skb) + nhoff);
+	if (!iph || iph->ihl * 4 < sizeof(struct iphdr)) {
 		pr_debug("IPv4 packet too short, accepted");
 		return NF_ACCEPT;
 	}
@@ -115,10 +93,7 @@ static unsigned int dp_mcast_br_hook_ipv6(struct sk_buff *skb, int nhoff)
 		return NF_ACCEPT;
 	}
 
-	if (nexthdr != NEXTHDR_ICMP)
-		return NF_ACCEPT;
-
-	if (dp_mcast_ipv6_is_mld_pkt(skb, nhoff) == 1) {
+	if (ipv6_is_mld(skb, nexthdr, nhoff)) {
 		pr_debug("IPv6 MLD packet dropped");
 		return NF_DROP;
 	}
